File: Problems/Leetcode/Leetcode/15_triplet_sum.py
class Solution:
    def threeSum(self, nums):
        nums.sort()
        res = []
        n = len(nums)

        for i in range(0, n):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j = i+1
            k = n-1
    
            while(j < k):
                if nums[i] + nums[j] + nums[k] == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1

                    while(j < k and nums[j] == nums[j-1]):
                        j += 1

                elif nums[i] + nums[j] + nums[k] < 0:
                    j += 1

                else:
                    k -= 1
        return res

File: Problems/Leetcode/Leetcode/15_triplet_sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int n = nums.size();

        for (int i = 0; i < n; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int j = i + 1;
            int k = n - 1;

            while (j < k)
            {
                if (nums[i] + nums[j] + nums[k] == 0)
                {
                    res.push_back(vector<int>{nums[i], nums[j], nums[k]});

                    j += 1;
                    while (j < k && nums[j] == nums[j - 1])
                        j++;
                }
                else if (nums[i] + nums[j] + nums[k] < 0)
                {
                    j++;
                }
                else
                {
                    k--;
                }
            }
        }
        return res;
    }
};

int main()
{
    vector<int> v{-1, 0, 1, 2, -1, -4};
    vector<vector<int>> s = Solution().threeSum(v);

    return 0;
}